core/src/test/java/org/elasticsearch/common/util/BigArraysTests.java

```diff
@@ -377,4 +379,53 @@ public void testMaxSizeExceededOnResize() throws Exception {
         }
     }
 
+    public void testBreakerTripsBeforeAllocatingArray() throws Exception {
+        final int maxSize = 1 << scaledRandomIntBetween(15, 22);
+        final int size = maxSize + randomIntBetween(1, 1 << 15);
+        for (String type : Arrays.asList("Byte", "Int", "Long", "Float", "Double", "Object")) {
+            HierarchyCircuitBreakerService hcbs = new HierarchyCircuitBreakerService(
+                Settings.builder()
+                    .put(REQUEST_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), maxSize, ByteSizeUnit.BYTES)
+                    .build(),
+                new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS));
+            BigArrays bigArrays = new BigArrays(null, hcbs, false).withCircuitBreaking();
+            Method create = BigArrays.class.getMethod("new" + type + "Array", long.class);
```

Comment:
Reflection? :(

What about instead of using an array of `Byte`, `Int`, etc which seems really fragile, we use a list of lambdas that invoke the real method. That way if there is ever a refactoring it will fail during compilation rather than at runtime.